RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,1126 @@
+5. F# users are typically concerned with this feature at the boundaries of their system so that they can flow non-null data into the inner parts of their system.
+6. The feature produces warnings by default, with different classes of warnings (nullable, non-nullable) offering opt-in/opt-out mechanisms.
+7. All existing F# projects compile with warnings turned off by default. Only new projects have warnings on by default. Compiling older F# code with newer F# code does not opt-in the older F# code.
+8. F# non-nullness is reasonably "strong and trustworthy" today for F#-defined types and routine F# coding. This includes the explicit use of `option` types to represent the absence of information. No compile-time guarantees today will be "lowered" to account for this feature.
+9. The F# compiler will strive to provide flow analysis such that common null-check scenarios guarantee null-safety when working with nullable reference types.
+10. This feature is useful for F# in other contexts, such as interoperation with `string` in JavaScript via [Fable](http://fable.io/).
+11. F# programmers may be able to start to phase out their use of `[<AllowNullLiteral>]`, `Unchecked.defaultof<_>`, and the `null` type constraint when they need ad-hoc nullability for reference types declared in F#.
+12. Syntax for the feature feels "baked in" to the type system, and should not be horribly unpleasant to use.
+
+## Detailed design
+[design]: #detailed-design
+
+### .NET Metadata
+
+To remain backwards-compatible with existing codebases and interoperate with other .NET languages that support nullability as a concept, F# will emit and respect .NET metadata that concerns to the following:
+
+* If a type is a nullable reference type (**NOTE:** C# has no notion of this, so an assembly-level marker would be pending discussion)
+* If an assembly has nullability as a concept (regardless of what it was compiled with)
+* If a generic type constraint is nullable
+* If a given method/function "handles null", such as `String.IsNullOrWhiteSpace`
+
+C# 8.0 will emit and respect metadata for these scenarios, with a well-known set of names for each attribute. These attributes will also be what F# uses, though the behavior of F# in the face of these attributes is not necessarily identical to how C# 8.0 behaves in the face of these attributes.
+
+#### Representing nullable types
+
+The following attribute will be used to represent a type that is marked as nullable:
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(
+        AttributeTargets.Class |
+        AttributeTargets.GenericParameter |
+        AttributeTargets.Event |
+        AttributeTargets.Field |
+        AttributeTargets.Property |
+        AttributeTargets.Parameter |
+        AttributeTargets.ReturnValue,
+        AllowMultiple = false)]
+    public sealed class NullableAttribute : Attribute
+    {
+        public NullableAttribute() { }
+        public NullableAttribute(bool[] b) { }
```

Comment:
What does the `bool[]` mean?  Is that used for nullability annotations for nested generic types?

Reply:
Not sure, actually. @jcouv ?